src/rust/index.rs

```diff
@@ -0,0 +1,413 @@
+use rocksdb::{Options, DB};
+use serde::{Deserialize, Serialize};
+use sourmash::_hash_murmur;
+use sourmash::cmd::ComputeParameters;
+use sourmash::collection::Collection;
+use sourmash::manifest::Manifest;
+use sourmash::signature::{Signature, SigsTrait};
+use sourmash::sketch::minhash::KmerMinHash;
+use sourmash::storage::{FSStorage, InnerStorage};
+use sourmash_plugin_branchwater::search_significance::{
+    compute_inverse_document_frequency, get_hash_frequencies, Normalization,
+};
+use sourmash_plugin_branchwater::utils::multicollection::SmallSignature;
+
+use crate::aminoacid::AminoAcidAmbiguity;
+use crate::encoding::{get_encoding_fn_from_moltype, get_hash_function_from_moltype};
+use crate::kmer_signature::{KmerInfo, KmerSignature, SerializableSignature};
+use crate::protein::Protein;
+use crate::uniprot::UniProtEntry;
+
+/// Statistics for k-mer frequency analysis
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ProteomeIndexKmerStats {
+    pub idf: HashMap<u64, f64>, // Inverse document frequency for each k-mer hashvalue
+    pub frequency: HashMap<u64, f64>, // Raw frequency for each k-mer hashvalue
+}
+
+// Represents the serializable state of ProteomeIndex
+#[derive(Serialize, Deserialize)]
+struct ProteomeIndexState {
+    // Store signatures instead of Collection
+    signatures: Vec<SerializableSignature>,
+    combined_minhash: KmerMinHash,
+    protein_info: HashMap<String, KmerSignature>,
+    moltype: String,
+    ksize: u32,
+    scaled: u32,
+}
+
+pub struct ProteomeIndex {
+    // RocksDB instance for persistent storage
+    db: DB,
+
+    // Collection of all signatures for searching
+    collection: Arc<Mutex<Collection>>,
+
+    // Combined minhash of all proteins for statistics
+    combined_minhash: Arc<Mutex<KmerMinHash>>,
+
+    // Map of signature md5 -> protein k-mer information
+    protein_info: Arc<Mutex<HashMap<String, Protein>>>,
+
+    // Amino acid ambiguity handler
+    aa_ambiguity: Arc<AminoAcidAmbiguity>,
+
+    // Protein encoding function
+    encoding_fn: fn(u8) -> u8,
+
+    // Add moltype field for serialization
+    moltype: String,
+
+    // Add ksize field for serialization
+    ksize: u32,
+
+    // Add scaled field for serialization
+    scaled: u32,
+
+    // Statistics for k-mer frequencies and IDF
+    stats: ProteomeIndexKmerStats,
+}
+
+impl ProteomeIndex {
+    pub fn new<P: AsRef<Path>>(path: P, ksize: u32, scaled: u32, moltype: &str) -> Result<Self> {
+        // Create RocksDB options
+        let mut opts = Options::default();
+        opts.create_if_missing(true);
+
+        // Open the database
+        let db = DB::open(&opts, path)?;
+
+        let hash_function = match get_hash_function_from_moltype(moltype) {
+            Ok(value) => value,
+            Err(value) => return value,
+        };
+
+        let encoding_fn = match get_encoding_fn_from_moltype(moltype) {
+            Ok(value) => value,
+            Err(value) => return value,
+        };
```

Comment:
```suggestion
        let hash_function = get_hash_function_from_moltype(moltype)?;

        let encoding_fn = get_encoding_fn_from_moltype(moltype)?;
```

See suggestion in encoding.rs

Reply:
accepting these will make the tests pass after accepting the `encoding.rs` changes